src/archive/sevenz.rs

```diff
@@ -150,7 +159,51 @@ where
         }
 
         Ok(true)
-    })?;
+    };
+
+    let password = password.as_ref().map(|p| p.as_ref());
+
+    match password {
+        Some(password) => sevenz_rust::decompress_with_extract_fn_and_password(
+            reader,
+            output_path,
+            sevenz_rust::Password::from(password.to_str().unwrap()),
+            entry_extract_fn,
+        )?,
+        None => sevenz_rust::decompress_with_extract_fn(reader, output_path, entry_extract_fn)?,
+    };
 
     Ok(count)
 }
+
+/// List contents of `archive_path`, returning a vector of archive entries
+pub fn list_archive(
+    archive_path: &Path,
+    password: Option<impl AsRef<[u8]>>,
+) -> impl Iterator<Item = crate::Result<FileInArchive>> {
+    let reader = fs::File::open(archive_path).unwrap();
+    let password = password.as_ref().map(|p| p.as_ref());
+
+    let mut files = Vec::new();
+
+    let entry_extract_fn = |entry: &SevenZArchiveEntry, _: &mut dyn Read, _: &PathBuf| {
+        files.push(Ok(FileInArchive {
+            path: entry.name().into(),
+            is_dir: entry.is_directory(),
+        }));
+        Ok(true)
+    };
+
+    match password {
+        Some(password) => sevenz_rust::decompress_with_extract_fn_and_password(
+            reader,
+            ".",
+            sevenz_rust::Password::from(password.to_str().unwrap()),
```

Comment:
Same case for utf-8 invalid. reuse code from that previous let-else (is there a better way?)